File: SWEA/D4/1861. 정사각형 방/정사각형 방.py
from collections import deque
def move(X, Y):
    d = [(0, 1), (0, -1), (1, 0), (-1, 0)]
    q = deque([(X, Y)])
    room = 1
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx = x + d[i][0]
            ny = y + d[i][1]
            if 0 <= nx < N and 0 <= ny < N:
                if arr[nx][ny] == arr[x][y] + 1:
                    q.append((nx, ny))
                    room += 1
    return room
T = int(input())
for tc in range(1, T+1):
    N = int(input())
    arr = [list(map(int, input().split())) for _ in range(N)]
    lst = []
    for i in range(N):
        for j in range(N):
            lst.append((arr[i][j], move(i, j)))
    lst.sort(key = lambda x: (x[1], -x[0]))
    num, cnt = lst.pop()
    print(f'#{tc} {num} {cnt}')
    

File: SWEA/D4/1861. 정사각형 방/정사각형 방.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Solution {

	static int[][] dir = { { 0, 1 }, { 0, -1 }, { 1, 0 }, { -1, 0 } };
	static int[][] room;
	static int N;
	static List<int[]> res;
	static Queue<int[]> que;

	private static int move(int X, int Y) {

		que = new ArrayDeque<>();
		int cnt = 1;
		que.offer(new int[] { X, Y });

		while (!que.isEmpty()) {
			int[] out = que.poll();
			int x = out[0];
			int y = out[1];

			for (int d = 0; d < 4; d++) {
				int nx = x + dir[d][0];
				int ny = y + dir[d][1];
				if (nx < 0 || nx >= N || ny < 0 || ny >= N)
					continue;
				if (room[nx][ny] != room[x][y] + 1)
					continue;
				que.offer(new int[] { nx, ny });
				cnt++;
			}
		}
		return cnt;
	}

	public static void main(String[] args) throws NumberFormatException, IOException {

		BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
		int T = Integer.parseInt(bf.readLine());

		for (int test_case = 1; test_case <= T; test_case++) {

			N = Integer.parseInt(bf.readLine());
			room = new int[N][N];
			res = new ArrayList<>();

			for (int i = 0; i < N; i++) {
				StringTokenizer st = new StringTokenizer(bf.readLine());
				for (int j = 0; j < N; j++) {
					room[i][j] = Integer.parseInt(st.nextToken());
				}
			}

			for (int i = 0; i < N; i++) {
				for (int j = 0; j < N; j++) {
					res.add(new int[] { room[i][j], move(i, j) });
				}
			}

			Collections.sort(res, new Comparator<int[]>() {

				@Override
				public int compare(int[] o1, int[] o2) {
					if (o1[1] == o2[1]) {
						return o1[0] - o2[0];
					}
					return o2[1] - o1[1];
				}
			});

			System.out.println("#" + test_case + " " + res.get(0)[0] + " " + res.get(0)[1]);
		}
	}
}
